AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift

```diff
@@ -88,69 +83,228 @@
+                delegatedAuthenticationSDKOutput: authenticationSDKOutput,
+                deleteDelegatedAuthenticationCredential: deleteDelegatedAuthenticationCredential
+            )
+            let encodedFingerprintResult = try AdyenCoder.encodeBase64(fingerprintResult)
+            return encodedFingerprintResult
+        }
+        
+        // MARK: - Delegated Authentication
+        
+        /// This method checks;
+        /// 1. if DA has been registered on the device
+        /// 2. shows an approval screen if it has been registered
+        /// else calls the completion with a failure.
+        private func startApprovalFlow(
+            _ delegatedAuthenticationInput: String,
+            cardType: CardType?,
+            cardNumber: String?,
+            completion: @escaping (Result<(daOutput: String, delete: Bool?), ApprovalFlowError>) -> Void
+        ) {
+            isDeviceRegistered(delegatedAuthenticationInput: delegatedAuthenticationInput) { [weak self] registered in
+                guard let self else { return }
+                if registered {
+                    showApprovalScreen(delegatedAuthenticationInput: delegatedAuthenticationInput,
+                                       cardType: cardType,
+                                       cardNumber: cardNumber,
+                                       completion: completion)
+                } else {
+                    // setting the state to attempt the registration flow.
+                    delegatedAuthenticationState.attemptRegistration = true
+                    completion(.failure(.deviceIsNotRegistered))
                 }
-            } catch {
-                didFail(with: error, completionHandler: completionHandler)
             }
-            
         }
         
-        internal func performDelegatedAuthentication(_ fingerprintToken: ThreeDS2Component.FingerprintToken,
-                                                     completion: @escaping (Result<String, DelegateAuthenticationError>) -> Void) {
-            guard let delegatedAuthenticationInput = fingerprintToken.delegatedAuthenticationSDKInput else {
-                completion(.failure(DelegateAuthenticationError.authenticationFailed(cause: nil)))
-                return
+        // MARK: Delegated Authentication Approval
+        
+        private enum ApprovalFlowError: Error {
+            case fallbackTo3ds
+            case deviceIsNotRegistered
+            case authenticationServiceFailed(underlyingError: Error)
+            case removeCredentialServiceError(underlyingError: Error)
+        }
+        
+        private func showApprovalScreen(
+            delegatedAuthenticationInput: String,
+            cardType: CardType?,
+            cardNumber: String?,
+            completion: @escaping (Result<(daOutput: String, delete: Bool?), ApprovalFlowError>
+            ) -> Void
+        ) {
+            presenter.showApprovalScreen(
```

Comment:
There is quite a lot of indentation going on. Do you think we could split it up in multiple methods?